File: src/commands/get.ts
import { Command } from 'commander';
import { getTurbopufferClient } from '../client.js';
import chalk from 'chalk';

export function createGetCommand(): Command {
  const get = new Command('get')
    .description('Get a document by ID from a namespace')
    .argument('<id>', 'Document ID to retrieve')
    .requiredOption('-n, --namespace <name>', 'Namespace to query')
    .option('-r, --region <region>', 'Override the region (e.g., aws-us-east-1, gcp-us-central1)')
    .action(async (id: string, options: { namespace: string; region?: string }) => {
      const client = getTurbopufferClient(options.region);
      const namespace = options.namespace;

      try {
        console.log(chalk.bold(`\nQuerying document with ID: ${id} from namespace: ${namespace}\n`));

        // Get namespace reference
        const ns = client.namespace(namespace);

        // Query with id filter
        const result = await ns.query({
          filters: ['id', 'Eq', id],
          top_k: 1,
          include_attributes: true
        });

        // Check if document was found
        if (!result.rows || result.rows.length === 0) {
          console.log(chalk.yellow('Document not found'));
          process.exit(1);
        }

        // Get the document
        const doc = result.rows[0];

        // Display document
        console.log(chalk.cyan('Document:'));
        console.log(JSON.stringify(doc, null, 2));

        console.log(chalk.gray(`\nQuery took ${result.performance.query_execution_ms.toFixed(2)}ms`));
      } catch (error) {
        console.error('Error:', error instanceof Error ? error.message : String(error));
        process.exit(1);
      }
    });

  return get;
}

File: src/tpuff/commands/get.py
"""Get command for tpuff CLI."""

import json
import sys

import click
from rich.console import Console

from tpuff.client import get_namespace
from tpuff.utils.debug import debug_log


console = Console()


@click.command("get")
@click.argument("id")
@click.option("-n", "--namespace", required=True, help="Namespace to query")
@click.option("-r", "--region", help="Override the region (e.g., aws-us-east-1, gcp-us-central1)")
@click.pass_context
def get(
    ctx: click.Context,
    id: str,
    namespace: str,
    region: str | None,
) -> None:
    """Get a document by ID from a namespace."""
    try:
        console.print(f"\n[bold]Querying document with ID: {id} from namespace: {namespace}[/bold]\n")

        # Get namespace reference
        ns = get_namespace(namespace, region)

        # Query with id filter
        query_params = {
            "filters": ["id", "Eq", id],
            "top_k": 1,
            "include_attributes": True,
        }

        # Debug: Log query parameters
        debug_log("Query Parameters", query_params)

        result = ns.query(**query_params)

        # Debug: Log API response
        debug_log("Query Response", result)

        # Extract rows from the response
        rows = result.rows if hasattr(result, "rows") else []

        # Check if document was found
        if not rows:
            console.print("[yellow]Document not found[/yellow]")
            sys.exit(1)

        # Get the document
        doc = rows[0]

        # Convert to dict for display
        if hasattr(doc, "model_dump"):
            doc_dict = doc.model_dump()
        elif hasattr(doc, "to_dict"):
            doc_dict = doc.to_dict()
        else:
            doc_dict = {"id": getattr(doc, "id", "N/A")}

        # Display document
        console.print("[cyan]Document:[/cyan]")
        console.print(json.dumps(doc_dict, indent=2, default=str))

        # Show performance info
        if hasattr(result, "performance") and result.performance:
            console.print(
                f"\n[dim]Query took {result.performance.query_execution_ms:.2f}ms[/dim]"
            )

    except Exception as e:
        console.print(f"[red]Error: {e}[/red]")
        sys.exit(1)
